### PR TITLE
test(parser): port over vLLM's tests (categories PARSER.batch.{2,4,5,6,7}.*), part 2

### PR DESCRIPTION
**Pull request by keivenchang**
#### Overview:

Follows PR #9338 (the batch.8 pilot, merged 2026-05-11). Extends the per-top-level-case sub-case rollout to the five remaining buckets where vLLM's test corpus organically partitions on a meaningful shape axis — `batch.2` (multi-call), `batch.4` (malformed), `batch.5` (missing end-token), `batch.6` (empty args), `batch.7` (complex types).

#### Details:

- **What lands here?** 293 new sub-case fixtures across 19 families × 5 buckets. Per-top-level-case file convention (`<family>/PARSER.batch.<n>.yaml`) extended bucket-by-bucket; legacy bare-key entries retired via orphan-rule.
- **Why these 5 buckets?** Audited in `~/dynamo/notes/VLLM_TEST_AUDIT.md` (private). Each carries genuine shape variation in vLLM's tests, not just parser-distribution. `batch.1`/`.3`/`.9`/`.10` were skipped — their row counts come from inherited common-suite repeating across N parsers, splitting would yield mostly-empty cells.
- **Per-bucket axes** (full taxonomy in `lib/parsers/PARSER_CASES.md`):
  - `batch.2`: `.a` parallel · `.b` multi-section back-to-back · `.c` surrounded with content · `.d` same-name twice (id distinctness)
  - `batch.4`: `.a` catch-all garbage · `.b` invalid JSON syntax · `.c` missing structural keys · `.d` malformed wrapper / mismatched tags
  - `batch.5`: `.a` missing closing tag · `.b` missing opening tag · `.c` truncation mid-call body
  - `batch.6`: `.a` canonical empty `{}` · `.b` zero-arg formatting · `.c` no-args-key
  - `batch.7`: `.a` standard scalar/container types · `.b` escaped/Unicode · `.c` type coercion · `.d` multi-arg/nested
- **Coverage gaps documented as skips:** pure-XML grammars (qwen3_coder, glm47, minimax_m2, deepseek_v4, deepseek_v3_2, nemotron_nano) skip sub-cases that need JSON arg structure (`batch.4.b`, `batch.6.c`, `batch.7.c/.d`). llama3_json skips `batch.5.b` (no paired close fence). harmony skips `batch.4.d` (no paired begin/end). These show as `n/a` in the matrix, not failures.
- **`KNOWN_DIVERGENCES` registry** restructured: 30+ stale bare-key entries dropped; ~80 sub-case entries added (vllm divergences from local pytest, sglang provisional `.a` entries inherited from prior bare-key set). One `EXPECTS_ERROR(KeyError):` prefix on `mistral.4.c` (vLLM's `mistral_tool_parser` raises on missing name key). Sglang `.b/.c/.d` divergences will surface when CI runs with sglang installed.
- **Tooling:** 2 new tool variants in `regenerate_fixtures.py` — `_BOOK_FLIGHT_MIXED` (string + int + bool for `batch.7.a`) and `_SET_TEMP_NUMERIC` (integer schema for `batch.7.c` coercion test).

#### Coverage matrix (mirrors `tests/parity/README.md`)

Each row is one parser (one wire-format family); the **Model(s)** column lists what model(s) that parser serves (many parsers cover multiple models — `mistral` covers the Mistral series, `hermes` covers Hermes-3 + various Llama variants, etc.). Columns are sub-case IDs from `PARSER_CASES.md`: top-level buckets (1, 3, 9, 10) have no sub-cases; the rest split into `.a`-`.d`.

Cell legend: `✓` = both vLLM and SGLang match Dynamo; `V` = vLLM diverges (xfailed); `S` = SGLang diverges; `VS` = both diverge; `n/a` = **not applicable** — no peer parser, OR the sub-case shape doesn't apply to this grammar (e.g. DSML families have no `4.b` because there's no embedded JSON to malform). Row-suffix `†` = no vLLM peer; `§` = no SGLang peer.

| model | parser | 1 | 2.a | 2.b | 2.c | 2.d | 3 | 4.a | 4.b | 4.c | 4.d | 5.a | 5.b | 5.c | 6.a | 6.b | 6.c | 7.a | 7.b | 7.c | 7.d | 8.a | 8.b | 8.c | 8.d | 9 | 10 |
|---|---|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
| **Top-N models** |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |
| DeepSeek V4 | deepseek_v4 § | ✓ | ✓ | V | V | ✓ | ✓ | V | n/a | V | ✓ | V | ✓ | V | ✓ | ✓ | n/a | V | ✓ | n/a | n/a | ✓ | V | V | V | ✓ | ✓ |
| Gemma 4 | gemma4 § | ✓ | ✓ | n/a | V | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | V | V | V | ✓ | ✓ |
| GLM 5.1 | glm47 | ✓ | ✓ | n/a | V | ✓ | ✓ | VS | n/a | n/a | ✓ | VS | ✓ | VS | ✓ | S | n/a | ✓ | ✓ | n/a | n/a | V | V | V | V | ✓ | ✓ |
| gpt-oss | harmony † | S | S | n/a | S | S | ✓ | S | S | ✓ | n/a | S | ✓ | S | S | S | ✓ | S | S | S | S | S | S | S | S | ✓ | S |
| Kimi K2.6 | kimi_k2 | ✓ | ✓ | ✓ | VS | ✓ | ✓ | ✓ | S | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | VS | VS | VS | VS | ✓ | ✓ |
| MiniMax 2.7 | minimax_m2 | ✓ | ✓ | S | S | ✓ | ✓ | V | n/a | V | V | VS | ✓ | VS | ✓ | ✓ | n/a | V | ✓ | n/a | n/a | ✓ | S | S | S | ✓ | ✓ |
| Nemotron (DeciLM) | nemotron_deci †§ | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a |
| Qwen 3 Coder | qwen3_coder | ✓ | ✓ | n/a | ✓ | ✓ | ✓ | V | n/a | n/a | ✓ | ✓ | V | ✓ | ✓ | ✓ | n/a | VS | ✓ | n/a | n/a | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ |
| **Others** |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |
| DeepSeek V3 | deepseek_v3 | ✓ | ✓ | ✓ | S | ✓ | ✓ | VS | V | VS | VS | VS | ✓ | VS | ✓ | V | V | ✓ | ✓ | ✓ | ✓ | S | ✓ | S | S | ✓ | ✓ |
| DeepSeek V3.1 | deepseek_v3_1 | ✓ | ✓ | ✓ | S | ✓ | ✓ | VS | V | VS | VS | VS | ✓ | VS | ✓ | ✓ | V | ✓ | ✓ | ✓ | ✓ | S | ✓ | S | S | ✓ | ✓ |
| DeepSeek V3.2 | deepseek_v3_2 | ✓ | ✓ | VS | VS | ✓ | ✓ | V | n/a | V | ✓ | V | ✓ | V | ✓ | ✓ | n/a | V | ✓ | n/a | n/a | S | VS | VS | VS | ✓ | ✓ |
| Hermes-3 / Llama variants | hermes | ✓ | ✓ | n/a | V | ✓ | ✓ | ✓ | VS | S | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | S | ✓ | ✓ | ✓ | ✓ | V | ✓ | V | V | ✓ | ✓ |
| AI21 Jamba | jamba § | ✓ | ✓ | n/a | V | ✓ | ✓ | ✓ | ✓ | V | V | V | ✓ | ✓ | ✓ | ✓ | V | ✓ | ✓ | ✓ | ✓ | V | ✓ | V | V | ✓ | ✓ |
| Llama 3 (JSON fmt) | llama3_json § | ✓ | V | n/a | V | V | ✓ | V | V | V | n/a | V | n/a | V | ✓ | ✓ | V | ✓ | ✓ | ✓ | ✓ | V | ✓ | V | V | ✓ | V |
| Mistral series | mistral | S | S | n/a | VS | S | ✓ | VS | VS | V | S | S | ✓ | VS | S | S | V | S | S | S | S | VS | S | VS | V | ✓ | S |
| Nemotron Nano | nemotron_nano †§ | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a | n/a |
| Phi-4 | phi4 § | ✓ | ✓ | n/a | V | ✓ | ✓ | ✓ | ✓ | V | V | V | V | V | ✓ | ✓ | V | ✓ | ✓ | ✓ | V | V | ✓ | V | V | ✓ | ✓ |
| Llama 4 (pythonic fmt) | pythonic | ✓ | ✓ | n/a | VS | ✓ | ✓ | ✓ | ✓ | n/a | ✓ | ✓ | ✓ | ✓ | ✓ | ✓ | n/a | ✓ | ✓ | ✓ | ✓ | VS | VS | VS | VS | ✓ | ✓ |
| Qwen 2.5 | qwen25 † | S | S | n/a | S | S | ✓ | ✓ | S | ✓ | S | S | ✓ | ✓ | S | S | ✓ | S | S | S | S | S | S | S | S | ✓ | S |

Each non-`✓` / non-`n/a` cell corresponds to an entry in `KNOWN_DIVERGENCES` at [`tests/parity/parser/test_parity_parser.py`](https://github.com/ai-dynamo/dynamo/blob/keivenchang/DIS-1936__parser-fixture-per-case-part-2/tests/parity/parser/test_parity_parser.py).

#### Verification:

693 passed (+307 vs pre-rollout baseline) / 545 skipped / 112 xfailed / 0 failed in local pytest with sglang skipped (not installed). All failures land as registered divergences via `KNOWN_DIVERGENCES`.

#### Where should the reviewer start?

`lib/parsers/PARSER_CASES.md` (the new sub-case taxonomy section under each top-level case), then `tests/parity/parser/regenerate_fixtures.py` INPUTS dict for the per-family shapes, then any one fixture file (e.g. `tests/parity/parser/fixtures/kimi_k2/PARSER.batch.4.yaml`) to see the embedded `TODO(research)` divergence comments inline. Divergence registry: `KNOWN_DIVERGENCES` in `tests/parity/parser/test_parity_parser.py`.

#### Related Issues:

Relates to [DIS-1936](https://linear.app/nvidia/issue/DIS-1936)
Builds on #9338 (merged 2026-05-11).

/coderabbit profile chill

## Summary by CodeRabbit

* **Documentation**
  * Expanded parser testing taxonomy with detailed sub-case breakdowns across batch-mode categories.
  * Updated parity harness documentation with refreshed test matrix layout and fixture provenance guidelines.

* **Tests**
  * Reorganized batch parser test fixtures across multiple parser implementations.
  * Added comprehensive test coverage for malformed inputs, edge cases, and argument parsing scenarios.
  * Updated divergence registry for cross-implementation parity assertions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-dynamo/dynamo/pull/9363)

### CODE REVIEW

**Review by keivenchang — COMMENTED** [review]
both fixed in ae8a36644f9.

batch.7.c: reframed as "schema mismatch — string value where schema declares typed primitive" with the contract pinned as *value-preservation* (Dynamo's behavior — keep `'20'` as string). Schema-coercing impls (vLLM's deepseek_v3_2) flip to int and surface as registered divergences. Renamed 13 fixture descriptions, the `PARSER_CASES.md` sub-case definition, and the matching `VLLM_TEST_AUDIT.md` rows. The fixtures' `expected` is unchanged — only the *label* is now consistent with what's actually asserted.

README staleness: cost row updated to "~5 s for 1350 tests" with the current pass/skip/xfail breakdown; the example fixture schema's `ref:` field now uses `ref: originated from <url>` per the convention.

Thanks for the structural pre-checks too — saved a separate audit pass.

**Review by keivenchang — COMMENTED** [review]
both fixed in ae8a36644f9.

batch.7.c: reframed as "schema mismatch — string value where schema declares typed primitive" with the contract pinned as *value-preservation* (Dynamos behavior — keep `'20'` as string). Schema-coercing impls (vLLMs deepseek_v3_2) flip to int and surface as registered divergences. Renamed 13 fixture descriptions, the `PARSER_CASES.md` sub-case definition, and the matching `VLLM_TEST_AUDIT.md` rows. The fixtures `expected` is unchanged — only the *label* is now consistent with what is actually asserted.

README staleness: cost row updated to "~5 s for 1350 tests" with current pass/skip/xfail breakdown; example fixture schema `ref:` field now uses `ref: originated from <url>` per the convention.

Thanks for the structural pre-checks too — saved a separate audit pass.

**Comment by keivenchang** [comment]
both fixed in ae8a36644f9.

batch.7.c: reframed as 'schema mismatch — string value where schema declares typed primitive' with contract pinned as *value-preservation* (Dynamo's behavior — keep `'20'` as string). Schema-coercing impls (vLLM's deepseek_v3_2) flip to int and surface as registered divergences. Renamed 13 fixture descriptions, the `PARSER_CASES.md` sub-case definition, and the matching `VLLM_TEST_AUDIT.md` rows. The fixtures' `expected` is unchanged — only the *label* is now consistent with what's actually asserted.

README staleness: cost row updated to '~5 s for 1350 tests' with current pass/skip/xfail breakdown; example fixture schema `ref:` field now uses `ref: originated from <url>` per the convention.

Thanks for the structural pre-checks too — saved a separate audit pass.

File: tests/parity/parser/fixtures/deepseek_v3_1/PARSER.batch.6.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    model_text: <｜tool▁calls▁begin｜><｜tool▁call▁begin｜>get_time<｜tool▁call▁end｜><｜tool▁calls▁end｜>
+    tools:
+    - *id001
+    expected:
+      calls: []
+      normal_text: <｜tool▁calls▁begin｜><｜tool▁call▁begin｜>get_time<｜tool▁call▁end｜><｜tool▁calls▁end｜>
+    # TODO(research): vllm diverges — vLLM rejects when tool_sep / args section is absent; Dynamo extracts the call
+    # vllm produces:
+    #   calls: []
+    #   normal_text: None
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/deepseek_v3_1/PARSER.batch.6.yaml]: fixed in 1365c5f — the divergence is on normal_text not call extraction (both impls return calls=[]); reason and TODO updated. @coderabbitai does this address the issue, and is this conversation resolvable? if so, please resolve conversation.

File: tests/parity/parser/fixtures/jamba/PARSER.batch.6.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    model_text: '<tool_calls>[{"name": "get_time"}]</tool_calls>'
+    tools:
+    - *id001
+    expected:
+      calls: []
+      normal_text: ''
+    # TODO(research): vllm diverges — vLLM accepts call with missing arguments key as having empty args; Dynamo rejects
+    # vllm produces:
+    #   calls: []
+    #   normal_text: '<tool_calls>[{"name": "get_time"}]</tool_calls>'
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/jamba/PARSER.batch.6.yaml]: fixed in 1365c5f — same shape across jamba/llama3_json/phi4: both impls return calls=[], divergence is normal_text (vLLM preserves raw, Dynamo emits empty). @coderabbitai does this address the issue, and is this conversation resolvable? if so, please resolve conversation.

File: tests/parity/parser/fixtures/phi4/PARSER.batch.6.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    model_text: 'functools[{"name": "get_time"}]'
+    tools:
+    - *id001
+    expected:
+      calls: []
+      normal_text: ''
+    # TODO(research): vllm diverges — vLLM accepts call with missing arguments key as having empty args; Dynamo rejects
+    # vllm produces:
+    #   calls: []
+    #   normal_text: 'functools[{"name": "get_time"}]'
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/phi4/PARSER.batch.6.yaml]: fixed in 1365c5f — phi4/6.c reason + TODO updated to describe the actual normal_text divergence (both impls return calls=[]). @coderabbitai does this address the issue, and is this conversation resolvable? if so, please resolve conversation.

File: tests/parity/parser/fixtures/deepseek_v3/PARSER.batch.2.yaml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      ```<｜tool▁call▁end｜><｜tool▁calls▁end｜>
+    tools:
+    - *id001
+    - *id001
+    expected:
+      calls:
+      - name: get_weather
+        arguments:
+          location: NYC
+      - name: get_weather
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/deepseek_v3/PARSER.batch.2.yaml:107]: not fixing — deepseek_v3 and deepseek_v4 are different parsers (v3 is fenced-JSON, v4 is DSML attribute-encoded); v3 doesn't diverge from vLLM on 2.c (no entry in KNOWN_DIVERGENCES), so no TODO is needed. The asymmetry with deepseek_v4 is intentional — TODOs appear only where divergences are registered. @coderabbitai does this address the issue, and is this conversation resolvable? if so, please resolve conversation.

File: tests/parity/parser/fixtures/nemotron_deci/PARSER.batch.2.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    description: Two back-to-back TOOLCALL wrappers
+    ref: dynamo
+    model_text: '<TOOLCALL>[{"name": "get_weather", "arguments": {"location": "NYC"}}]</TOOLCALL><TOOLCALL>[{"name": "get_time",
+      "arguments": {"timezone": "EST"}}]</TOOLCALL>'
+    tools:
+    - *id001
+    - *id002
+    expected:
+      calls: []
+      normal_text: ''
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/nemotron_deci/PARSER.batch.2.yaml:46]: not adjusting expected in 1365c5f — fixture is regenerated from Dynamo as oracle (see regenerate_fixtures.py); pinning a non-empty calls value would diverge from actual output and fail the test. Added an inline NOTE explaining the parser limitation (no restart on back-to-back wrappers). nemotron_deci is n/a in the parity matrix (no vLLM/SGLang peer) so this is Dynamo self-check only — parser fix is out of scope here. @coderabbitai does this address the issue, and is this conversation resolvable? if so, please resolve conversation.

File: tests/parity/parser/fixtures/nemotron_deci/PARSER.batch.2.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
```
[the remark this answers is not in the document]
Comment (keivenchang) [reply tests/parity/parser/fixtures/nemotron_deci/PARSER.batch.2.yaml:46]: already addressed in 1365c5f — added inline NOTE that nemotron_deci's parser doesn't restart on back-to-back wrappers (Dynamo self-check only, n/a row in parity matrix); fixture pins actual oracle output. See tests/parity/parser/fixtures/nemotron_deci/PARSER.batch.2.yaml:47-52.

**Review by KrishnanPrash — APPROVED** [review]

**Comment by keivenchang** [comment]
@coderabbitai, if addressed please Resolve conversation.

File: components/src/dynamo/global_planner/tests/unit/test_main_endpoint_registration.py
Comment (rmccorm4) [inline components/src/dynamo/global_planner/tests/unit/test_main_endpoint_registration.py:1]: @keivenchang why are all these `planner` related changes in this PR? Can we remove them?

Comment (keivenchang) [reply components/src/dynamo/global_planner/tests/unit/test_main_endpoint_registration.py:1]: hey Ryan, you're right — they got absorbed by accident during a soft-reset rebase a while back. cleaned up in a4de7c8, just force-pushed. PR diff is now only tests/parity/parser/**, tests/parity/README.md, and lib/parsers/PARSER_CASES.md (130 files, all parser test fixtures + the taxonomy doc).

**Comment by rmccorm4** [comment]
Failing CPU test should be fixed here soon: https://github.com/ai-dynamo/dynamo/pull/9463
